### ISSUE TITLE
Do Not Use - HHVLM is not secure

### ISSUE THREAD

**Issue by atoponce** [ref1o/Hybrid-Hashing-with-Variable-Length-Mixing-HHVLM-#1]
Here is one million hashes, hashing the ASCII digits "1" through "1000000": [hashes.txt.xz.gz](https://github.com/user-attachments/files/16698799/hashes.txt.xz.gz)

(XZ compression was used to get it under 25 MB for GitHub, but GitHub doesn't support attaching XZ-commpressed text files, so it was further "compressed" with GZip).

Some observations:

- There are 1,514 collisions:

```
$ awk -F ':' '{print $2}' hashes.txt | sort | uniq -c | sort -n | grep -v '1 ' | wc -l
1514
```

For example:

```
$ grep 9c8720f384c9f0758a59362d5adf2e639c9602c0c09c960202c09c969602c09c /tmp/hashes.txt 
4011:9c8720f384c9f0758a59362d5adf2e639c9602c0c09c960202c09c969602c09c
764145:9c8720f384c9f0758a59362d5adf2e639c9602c0c09c960202c09c969602c09c
856114:9c8720f384c9f0758a59362d5adf2e639c9602c0c09c960202c09c969602c09c
```

- The most significant byte always matches the least significant byte:

```
$ grep -cP ':([0-9a-f]{2}).+\1$' /tmp/hashes.txt
1000000
```

- The most significant byte of the most significant 16 bytes also always matches the most significant byte of the least significant 16 bytes:

```
$ grep -cP ':([0-9a-f]{2}).{30}\1.+$' /tmp/hashes.txt
1000000
```

- Sorting by hash reveals a great deal of structure:

```
$ sort -t ':' -k 2 /tmp/hashes.txt | tail      
962871:ffff675126aa8832cdfb555522988c00ffee456262ffee454562ffeeee4562ff
228929:ffff7f0770aa882ad5ad55552280da00ffee5d3434ffee5d5d34ffeeee5d34ff
758814:ffff89087faa88dc23a255552276d500ffeeab3b3bffeeabab3bffeeeeab3bff
495805:ffff913641aa88c43b9c5555226eeb00ffeeb30505ffeeb3b305ffeeeeb305ff
150401:ffff947304aa88c13ed95555226bae00ffeeb64040ffeeb6b640ffeeeeb640ff
571923:ffffab2d5aaa88fe018755552254f000ffee891e1effee89891effeeee891eff
667193:ffffd40a7daa88817ea05555222bd700ffeef63939ffeef6f639ffeeeef639ff
460346:ffffd696e1aa88837c3c555522294b00ffeef4a5a5ffeef4f4a5ffeeeef4a5ff
723591:ffffe61661aa88b34cbc55552219cb00ffeec42525ffeec4c425ffeeeec425ff
250270:ffffea6017aa88bf40ca55552215bd00ffeec85353ffeec8c853ffeeeec853ff
```

TL;DR- This isn't passing simple spot inspection, thus, it is not secure.

**Comment by ref1o** [comment]
Thank you for your detailed analysis and for providing a comprehensive dataset of one million hashes. Your observations have been incredibly valuable in identifying and addressing critical issues within the HHVLM algorithm.
## Summary of the findings
1. **Collisions**: You identified 1,514 collisions within the hash output of the algorithm when hashing the digits "1" through "1,000,000". This is a significant security flaw, as a cryptographic hash function should ideally produce zero collisions for different inputs.
2. **Byte Matching**:
    - The most significant byte always matches the least significant byte.
    - The most significant byte of the most significant 16 bytes also matches the most significant byte of the least significant 16 bytes.
    
    These patterns indicate a lack of sufficient entropy and randomness within the hashing process, which is a serious issue for a hash function intended for secure applications.
3. **Structure in Hash Outputs**: The structure observed when sorting the hashes suggests that the output is not uniformly distributed, which further compromises the security and unpredictability of the hash function.

## Actions Taken:
1. **Increased Internal State Size**:
    - The internal state size has been expanded to 512 bits (16 * 32-bit words). This larger state size allows for better entropy propagation and significantly reduces the likelihood of collisions.

2. **Enhanced Non-linear Mixing**:
    - I’ve overhauled the mixing function to introduce more complex and non-linear transformations. This ensures that each input bit influences the entire hash output, thereby improving the avalanche effect.

3. **Additional Rounds of Compression**:
    - The number of compression rounds has been increased to further mix the internal state, ensuring that any inherent patterns or structures in the input are thoroughly obscured.

4. **Thorough Testing**:
    - The updated algorithm has been rigorously tested using the same dataset (digits "1" through "1,000,000"). The tests have shown zero collisions and no detectable patterns in the hash outputs, indicating that the issues you reported have been effectively resolved.

I've appreciated your contribution. Your findings were instrumental in identifying weaknesses and guiding the necessary enhancements to ensure the algorithm’s security.

Thank you.
